internal/sidekick/internal/parser/discovery/uritemplate.go

```diff
@@ -0,0 +1,123 @@
+// Copyright 2025 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package discovery
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/googleapis/librarian/internal/sidekick/internal/api"
+)
+
+const (
+	beginExpression = '{'
+	endExpression   = '}'
+	slash           = '/'
+)
+
+var (
+	identifierRe = regexp.MustCompile("[A-Za-z][A-Za-z0-9_]*")
+)
+
+// ParseUriTemplate parses a [RFC 6570] URI template as an `api.PathTemplate`.
+//
+// In sidekick we need to capture the structure of the URI template for the
+// codec(s) to emit good templates with them.
+func ParseUriTemplate(uriTemplate string) (*api.PathTemplate, error) {
+	template := &api.PathTemplate{}
+	var pos int
+	for {
+		var err error
+		var segment *api.PathSegment
+		var width int
+
+		if pos == len(uriTemplate) {
+			return nil, fmt.Errorf("expected a segment, found eof: %s", uriTemplate)
+		}
+		if uriTemplate[pos] == beginExpression {
+			segment, width, err = parseExpression(uriTemplate[pos:])
+		} else {
+			segment, width, err = parseLiteral(uriTemplate[pos:])
+		}
+		if err != nil {
+			return nil, err
+		}
+		template.Segments = append(template.Segments, *segment)
+		pos += width
+		if pos == len(uriTemplate) || uriTemplate[pos] != slash {
+			break
+		}
+		pos++ // Skip slash
+	}
+	return template, nil
+}
+
+func parseExpression(input string) (*api.PathSegment, int, error) {
+	if input == "" || input[0] != beginExpression {
+		return nil, 0, fmt.Errorf("missing `{` character in expression %q", input)
+	}
+	tail := input[1:]
+	if strings.IndexAny(tail, "+#") == 0 {
+		return nil, 0, fmt.Errorf("level 2 expressions unsupposed input=%q", input)
+	}
+	if strings.IndexAny(tail, "./?&") == 0 {
+		return nil, 0, fmt.Errorf("level 3 expressions unsupposed input=%q", input)
+	}
+	if strings.IndexAny(tail, "=,!@|") == 0 {
+		return nil, 0, fmt.Errorf("reserved character on expression %q", input)
+	}
+	match := identifierRe.FindStringIndex(tail)
+	if match[0] != 0 {
+		return nil, 0, fmt.Errorf("no identifier found on expression %q", input)
+	}
+	id := tail[0:match[1]]
+	tail = tail[match[1]:]
+	if tail == "" || tail[0] != endExpression {
```

Comment:
should we also check if `tail[1]` is a slash. We do that in `parseLiteral`.

Would the parser incorrectly accept input like `{foo}}bar` ?